hazelcast/src/main/java/com/hazelcast/partition/PartitionAware.java

```diff
@@ -37,6 +37,15 @@
  * {@link com.hazelcast.core.DistributedObject} also has a notion of the partition key which is of type String
  * to ensure that the same partition as distributed Objects Strings is used for the partition key.
  *
+ * Additionally, it is expected that partition key remains same for a given object throughout its lifetime in the
```

Comment:
I propose to use similar form as Java uses for documenting such contracts, eg. for `equals` and `hashCode`, as mathematical axioms/definitions. Examples and rationale may follow.

AFAIR we have the following important properties:
1. for equal (TODO: Object.equals or serialized form equality?) objects partitionKey is the same
2. partitionKey of given object does not change over time (to a degree implied by 1 but worth specifying explicitly)

Reply:
also those properties must be true for entire time in entire cluster (so they apply also to user code evolution if they want to change implementation of `PartitionAware` for their objects) and for persistent IMap with PartitionAware keys maybe as long as such IMap exists.